Storage/src/StorageObject.php

```diff
@@ -672,13 +676,59 @@ public function downloadToFile($path, array $options = [])
      */
     public function downloadAsStream(array $options = [])
     {
-        return $this->connection->downloadObject(
+        $headers = $options['restOptions']['headers'] ?? [];
+        $resultStream = Utils::streamFor(null);
+        $userSuppliedRangeHeaders = isset($headers['Range']);
+        $expectedSize = isset($options['size']) ? $options['size'] : $this->info()['size'];
+
+        // add Range headers, which default to the whole object
+        // but only if the request doesn't already contain any range headers
+        if (!$userSuppliedRangeHeaders) {
+            $headers['Range'] = sprintf("bytes=%s-", $resultStream->getSize());
+            $options = array_merge($options, ['restOptions'=>['headers'=>$headers]]);
+
+            $options += [
+                'onRetryException' => function (\Exception $e, $attempt, &$arguments) use ($resultStream) {
+                    // if the exception has a response for us to use
+                    if ($e instanceof RequestException && $e->hasResponse()) {
+                        $msg = (string) $e->getResponse()->getBody();
+
+                        $fetchedStream = Utils::streamFor($msg);
+
+                        // add the partial response to our stream that we will return
+                        Utils::copyToStream($fetchedStream, $resultStream);
+
+                        // modify the range headers to fetch the remaining data
+                        $arguments[1]['headers']['Range'] = sprintf('bytes=%s-', $resultStream->getSize());
+                    }
+                }
+            ];
+        }
+
+        $fetchedStream = $this->connection->downloadObject(
             $this->formatEncryptionHeaders(
                 $options
                 + $this->encryptionData
                 + array_filter($this->identity)
             )
         );
+
+        // There are 2 cases when we want to return the fetched stream
+        // and not merge it with a partially fetched stream previously.
+        // 1. If the user supplied their own Range headers
+        // 2. If the object had transcoding enabled, the fetched stream
```

Comment:
In this case we should fast-forward the returned stream up to the last offset that was fed into the user stream.

Reply:
I'm not sure I follow. Could you elaborate with an example?

Reply:
Resolved offline :)